cert/spec.go

```diff
@@ -483,16 +475,56 @@ func (spec *Spec) TakeAction(changeType string) error {
 // The maximum number of attempts before giving up.
 const maxAttempts = 5
 
-// renewPKI Try to update the on disk PKI content with a fresh CA/cert as needed
-func (spec *Spec) renewPKI(ca *x509.Certificate) error {
-	metrics.SpecRefreshCount.WithLabelValues(spec.Path).Inc()
-	failed := true
+// writePKIToDisk writes the given PKI materials to disk, returning errors if anything occurs.
+// CA must always be passed, but is optionally written if the spec defines an on disk CA.
+// keypair can be nil if the spec does not write a cert/key- if it's just used for tracking the CA.
+func (spec *Spec) writePKIToDisk(ca *x509.Certificate, keyPair *tls.Certificate) (err error) {
+
+	metrics.SpecWriteCount.WithLabelValues(spec.Path).Inc()
+
 	defer func() {
-		if failed {
+		if err != nil {
 			metrics.SpecWriteFailureCount.WithLabelValues(spec.Path).Inc()
 		}
 	}()
 
+	if spec.CA.File != nil {
+		err = spec.CA.File.WriteCertificate(ca)
+		if err != nil {
+			return
+		}
+		spec.updateCAExpiry(ca.NotAfter)
+	}
+	if keyPair == nil {
+		return nil
+	}
+	keyData, err := encodeKeyToPem(keyPair.PrivateKey)
+	if err != nil {
+		return
+	}
+
+	err = spec.Cert.WriteCertificate(keyPair.Leaf)
+	if err != nil {
+		log.Errorf("spec %s: failed to write certificate to disk: %s", spec, err)
+		return
+	}
+	err = spec.Key.WriteFile(keyData)
+
+	if err != nil {
+		log.Errorf("spec %s: failed to write key to disk: %s", spec, err)
+		return
+	}
+	spec.updateCertExpiry(keyPair.Leaf.NotAfter)
+
+	return
+}
+
+// fetchNewKeyPair request a fresh certificate/key from the transport, backing off as needed.
+func (spec *Spec) fetchNewKeyPair() (*tls.Certificate, error) {
+	metrics.SpecRefreshCount.WithLabelValues(spec.Path).Inc()
+
+	defer spec.tr.Backoff.Reset()
```

Comment:
nit: Probably out of scope, but `tr` bugs me as a name, as it doesn't clearly describe whatever it is. I'll have to go look to see what `tr` is :)